.github/workflows/build.yml

```diff
@@ -12,48 +12,94 @@ concurrency:
 
 jobs:
   build:
-    name: "Build ${{ matrix.platform }} in ${{ matrix.build_type }}"
+    name: "Build ${{ matrix.platform.runtime }} in ${{ matrix.build_type }}"
     strategy:
       matrix:
-        platform: [windows, ubuntu, macos]
+        platform: [{ os: windows, runtime: windows-latest }, { os: ubuntu, runtime: ubuntu-latest }, { os: macos, runtime: macos-12 }]
         build_type: [Debug, Release]
     env:
       PARALLEL: -j 2
```

Comment:
Maybe, instead of duplicating most of this file, you could add macos-14 to the matrix and just have a condition here like
`if: ${{ github.ref == 'refs/heads/main' || matrix.platform.runtime != 'macos-14' }}`